### PR TITLE
Add IOperation support for string interpolation

### LINKED ISSUE #18300: IOperation doesn't support string interpolations
**Example Code** 
using System;
class P
{
    private void M()
    {
        var s = "";
        if (int.TryParse(s, out var i))
            Console.WriteLine($"i ={ i}, s ={ s}");
        else
          Console.WriteLine($"i ={ i}, s ={ s}");
    }
}

Try generating Ioperation tree for Ifstatement
**Generated Tree**

IOperation tree for "if (int.TryParse(s, out var i))
            Console.WriteLine($"i ={ i}, s ={ s}");
        else
          Console.WriteLine($"i ={ i}, s ={ s}");"

IIfStatement (OperationKind.IfStatement)
  Condition: IInvocationExpression (static System.Boolean System.Int32.TryParse(System.String s, out System.Int32 result)) (OperationKind.InvocationExpression, Type: System.Boolean)
      IArgument (Matching Parameter: s) (OperationKind.Argument)
        ILocalReferenceExpression: s (OperationKind.LocalReferenceExpression, Type: System.String)
      IArgument (Matching Parameter: result) (OperationKind.Argument)
        ILocalReferenceExpression: i (OperationKind.LocalReferenceExpression, Type: System.Int32)
  IExpressionStatement (OperationKind.ExpressionStatement)
    IInvocationExpression (static void System.Console.WriteLine(System.String value)) (OperationKind.InvocationExpression, Type: System.Void)
      IArgument (Matching Parameter: value) (OperationKind.Argument)
        IOperation:  (OperationKind.None)
  IExpressionStatement (OperationKind.ExpressionStatement)
    IInvocationExpression (static void System.Console.WriteLine(System.String value)) (OperationKind.InvocationExpression, Type: System.Void)
      IArgument (Matching Parameter: value) (OperationKind.Argument)
        IOperation:  (OperationKind.None)

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #18300 

```
public interface IInterpolatedStringExpression : IOperation
{
    /// <summary>
    /// Constituent parts of interpolated string.
    /// </summary>
    ImmutableArray<IInterpolatedStringContent> Parts { get; }
}

public interface IInterpolatedStringContent : IOperation
{
}

public interface IInterpolatedStringText : IInterpolatedStringContent
{
    /// <summary>
    /// Text content
    /// </summary>
    IOperation Text { get; }
}

public interface IInterpolation : IInterpolatedStringContent
{
    /// <summary>
    /// Expression
    /// </summary>
    IOperation Expression { get; }

    /// <summary>
    /// Optional alignment
    /// </summary>
    IOperation Alignment { get; }

    /// <summary>
    /// Optional format string.
    /// </summary>
    IOperation FormatString { get; }
}
```

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @dotnet/analyzer-ioperation @AlekseyTs @cston @dotnet/roslyn-compiler for review.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 using System.Linq;
 using System.Runtime.CompilerServices;
 using Microsoft.CodeAnalysis.CSharp;
+using Roslyn.Utilities;
```
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: @heejaechang Can you please review the Operation factory changes..

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 2]
```diff
@@ -165,6 +166,10 @@ private static IOperation CreateInternal(BoundNode boundNode)
                     return CreateBoundLabeledStatementOperation((BoundLabeledStatement)boundNode);
                 case BoundKind.ExpressionStatement:
                     return CreateBoundExpressionStatementOperation((BoundExpressionStatement)boundNode);
+                case BoundKind.InterpolatedString:
+                    return CreateBoundInterpolatedStringExpressionOperation((BoundInterpolatedString)boundNode);
+                case BoundKind.StringInsert:
+                    throw ExceptionUtilities.UnexpectedValue(boundNode.Kind);
```
Comment (heejaechang) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: rather than returning OperationKind.None we want to throw? I am fine with throwing, just want to see what is our principle is.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 3]
```diff
@@ -985,5 +990,62 @@ private static IExpressionStatement CreateBoundExpressionStatementOperation(Boun
...
+        {
+            Lazy<ImmutableArray<IInterpolatedStringContent>> parts = new Lazy<ImmutableArray<IInterpolatedStringContent>>(() => GetInterpolatedStringExpressionParts(boundInterpolatedString));
+            bool isInvalid = boundInterpolatedString.HasErrors;
+            SyntaxNode syntax = boundInterpolatedString.Syntax;
+            ITypeSymbol type = boundInterpolatedString.Type;
+            Optional<object> constantValue = ConvertToOptional(boundInterpolatedString.ConstantValue);
+            return new LazyInterpolatedStringExpression(parts, isInvalid, syntax, type, constantValue);
+        }
+
+        private static IInterpolatedStringContent CreateBoundInterpolatedStringContentOperation(BoundNode boundNode)
```
Comment (heejaechang) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: there is another file, Factory_Method.cs

that file is supposed to be where this kind of method should go in.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 2 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Binder only generates a BoundStringInsert node within BindInterpolatedString, and we handle  BoundKind.StringInsert in the newly added method below. Basically this validates the binder contract, we can have BoundStringInsert in the bind tree only under a BoundInterpolatedStringExpression.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 4]
```diff
@@ -985,5 +990,62 @@ private static IExpressionStatement CreateBoundExpressionStatementOperation(Boun
...
+        {
+            if (boundNode == null)
+            {
+                return null;
+            }
```
Comment (heejaechang) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: hmm... is there a reason why this diverge from existing pattern? the s_cache should never be used except in Create method.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 3 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: This method is dealing with actual bound nodes with a specific bound kind. I though the _Method.cs was only for helper methods, and this file for methods that generate Operation node from bound nodes. Is that not true?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 5]
```diff
@@ -985,5 +990,62 @@ private static IExpressionStatement CreateBoundExpressionStatementOperation(Boun
...
+            bool isInvalid = boundStringInsert.HasErrors;
+            SyntaxNode syntax = boundStringInsert.Syntax;
+            ITypeSymbol type = null;
+            Optional<object> constantValue = default(Optional<object>);
+            return new LazyInterpolation(expression, alignment, format, isInvalid, syntax, type, constantValue);
+        }
+
+        private static IInterpolatedStringText CreateBoundInterpolatedStringTextOperation(BoundLiteral boundLiteral)
+        {
+            Lazy<IOperation> text = new Lazy<IOperation>(() => CreateInternal(boundLiteral));
```
Comment (heejaechang) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: why is it calling CreateInternal?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 4 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Yes, the reason is a BoundLiteralExpression under a BoundInterpolatedExpressionString must map to an InterpolatedStringText operation node, not the common LiteralExpression node for strings seen at other places in the bound tree. Basically child bound nodes of an interpolated string are special. Do you see any other way to avoid this?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 6]
```diff
@@ -985,5 +990,62 @@ private static IExpressionStatement CreateBoundExpressionStatementOperation(Boun
...
+
+            return (IInterpolatedStringContent)s_cache.GetValue(boundNode, n => CreateBoundInterpolatedStringContentCore(n));
+        }
+
+        private static IInterpolatedStringContent CreateBoundInterpolatedStringContentCore(BoundNode boundInterpolatedStringContent)
+        {
+            switch (boundInterpolatedStringContent.Kind)
+            {
+                case BoundKind.StringInsert:
+                    return CreateBoundInterpolationOperation((BoundStringInsert)boundInterpolatedStringContent);
```
Comment (heejaechang) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: why are these not just call Create?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+                //         Console.WriteLine(/*<bind>*/$"String {x1} and constant {Class}"/*</bind>*/);
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "x1").WithArguments("x1").WithLocation(8, 47),
+                // CS0119: 'Class' is a type, which is not valid in the given context
+                //         Console.WriteLine(/*<bind>*/$"String {x1} and constant {Class}"/*</bind>*/);
+                Diagnostic(ErrorCode.ERR_BadSKunknown, "Class").WithArguments("Class", "type").WithLocation(8, 65)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<InterpolatedStringExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+    }
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs]: Could you add a test for an empty interpolated section? ie `$"{}"`?

File: src/Compilers/Core/Portable/Operations/IInterpolatedStringContent.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+{
+    /// <summary>
+    /// Represents a constituent part of an interpolated string.
+    /// </summary>
+    /// <remarks>
+    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
+    /// change it in the future.
+    /// </remarks>
+    public interface IInterpolatedStringContent : IOperation
+    {
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/Operations/IInterpolatedStringContent.cs]: no way to get content?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 5 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: The special handling is needed because BoundLiteralExpression under a BoundInterpolatedStringExpression is special. If we invoke Create method here, then it child of InterpolatedStringText node to be the same InterpolatedStringText node and walker goes into an infinite recursion. Let me stop by to explain it.

File: src/Compilers/Core/Portable/Operations/IInterpolatedStringContent.cs (hunk 8 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/Operations/IInterpolatedStringContent.cs]: ah. I see. sorry. ignore.

Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/IInterpolatedStringContent.cs]: Content is based on the sub-types of IInterpolatedStringContent (string versus interpolated expression)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 6 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: As we discussed offline, I will refactor this a bit to follow current pattern.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs (hunk 7 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs]: Added.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs [hunk 9]
```diff
@@ -2004,6 +2004,205 @@ public override void Accept(OperationVisitor visitor)
     }
 
     /// <remarks>
+    /// Represents an interpolated string expression.
+    /// </remarks>
+    internal abstract partial class BaseInterpolatedStringExpression : Operation, IInterpolatedStringExpression
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: >internal abstract partial class BaseInterpolatedStringExpression : Operation, IInterpolatedStringExpression [](start = 4, length = 107)

Is this file really generated? I do not see changes to Operations.xml in this PR.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 10]
```diff
@@ -985,5 +990,49 @@ private static IExpressionStatement CreateBoundExpressionStatementOperation(Boun
...
             return new LazyExpressionStatement(expression, isInvalid, syntax, type, constantValue);
         }
+
+        private static IInterpolatedStringExpression CreateBoundInterpolatedStringExpressionOperation(BoundInterpolatedString boundInterpolatedString)
+        {
+            Lazy<ImmutableArray<IInterpolatedStringContent>> parts = new Lazy<ImmutableArray<IInterpolatedStringContent>>(() => GetInterpolatedStringExpressionParts(boundInterpolatedString));
+            bool isInvalid = boundInterpolatedString.HasErrors;
+            SyntaxNode syntax = boundInterpolatedString.Syntax;
+            ITypeSymbol type = boundInterpolatedString.Type;
+            Optional<object> constantValue = ConvertToOptional(boundInterpolatedString.ConstantValue);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >Optional<object> constantValue = ConvertToOptional(boundInterpolatedString.ConstantValue); [](start = 12, length = 90)

Is there a good reason for making this optional even though the value is readily available?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 11]
```diff
@@ -985,5 +990,49 @@ private static IExpressionStatement CreateBoundExpressionStatementOperation(Boun
...
+            return new LazyInterpolatedStringExpression(parts, isInvalid, syntax, type, constantValue);
+        }
+
+        private static IInterpolatedStringContent CreateBoundInterpolatedStringContentOperation(BoundNode boundNode)
+        {
+            if (boundNode.Kind == BoundKind.Literal)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >boundNode.Kind == BoundKind.Literal [](start = 16, length = 35)

Should we also check that this is a string?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Given the shape of the API, I would check whether this is a BoundStringInsert instead

---
In reply to: [117335873](https://github.com/dotnet/roslyn/pull/19590#discussion_r117335873) [](ancestors = 117335873)

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 12]
```diff
@@ -1016,6 +1020,43 @@ Namespace Microsoft.CodeAnalysis.Semantics
...
+            Dim parts As New Lazy(Of ImmutableArray(Of IInterpolatedStringContent))(Function() GetInterpolatedStringExpressionParts(boundInterpolatedString))
+            Dim isInvalid As Boolean = boundInterpolatedString.HasErrors
+            Dim syntax As SyntaxNode = boundInterpolatedString.Syntax
+            Dim type As ITypeSymbol = boundInterpolatedString.Type
+            Dim constantValue As [Optional](Of Object) = ConvertToOptional(boundInterpolatedString.ConstantValueOpt)
+            Return New LazyInterpolatedStringExpression(parts, isInvalid, syntax, type, constantValue)
+        End Function
+
+        Private Shared Function CreateBoundInterpolatedStringContentOperation(boundNode As BoundNode) As IInterpolatedStringContent
+            If boundNode.Kind = BoundKind.Literal Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >If boundNode.Kind = BoundKind.Literal Then [](start = 12, length = 42)

Same comments as for C#.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+internal class Class
+{
+    public void M()
+    {
+        Console.WriteLine(/*<bind>*/$""""/*</bind>*/);
+    }
+}
+";
+            string expectedOperationTree = @"
+IInterpolatedStringExpression (OperationKind.InterpolatedStringExpression, Type: System.String) (Syntax: '$""""')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs]: There are no parts? This doesn't look right.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+    IInterpolation (OperationKind.Interpolation) (Syntax: '{1}')
+      Expression: ILiteralExpression (Text: 1) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1) (Syntax: '1')
+";
+            var expectedDiagnostics = DiagnosticDescription.None;
+
+            VerifyOperationTreeAndDiagnosticsForTest<InterpolatedStringExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [Fact, WorkItem(18300, "https://github.com/dotnet/roslyn/issues/18300")]
+        public void InterpolatedStringExpression_FormatAndAlignment()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs]: >InterpolatedStringExpression_FormatAndAlignment() [](start = 20, length = 49)

It doesn't look like we have a test for a scenario when all three members of IInterpolation are present.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs (hunk 13 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs]: We should probably open a design issue to confirm that this is what we expect.

---
In reply to: [117338430](https://github.com/dotnet/roslyn/pull/19590#discussion_r117338430) [](ancestors = 117338430)

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs (hunk 9 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: generator is not in yet. working on something else. I will put generator later, but named it as generated since I didn't want people to put random one in this file.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 10 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: I have followed the existing pattern in this type. Tagging @heejaechang

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 11 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: @AlekseyTs Are you recommending checking for both BoundLiteral (with type string) and BoundStringInsert in this method? Or are you just recommending we add an assert if we receive a bound node other than BoundLiteral/BoundStringInsert here?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs (hunk 13 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs]: Filed https://github.com/dotnet/roslyn/issues/19630

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs (hunk 14 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs]: Sure, will add a test

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 10 above)
Comment (heejaechang) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Constantvalue uses same pattern for all Operations. but unlike BoundNode which uses null as no value, Operation uses Optional to be more explicit.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 11 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: I would check only for BoundStringInsert because only BoundStringInsert can "fit" into IInterpolation, everything else can "fit" into the "text" part.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs (hunk 14 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IInterpolatedStringExpression.cs]: Added

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

**Comment by mavasani** [comment]
@cston @jcouv - can one of you please take a look?

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_unit32_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 15]
```diff
@@ -80,6 +80,9 @@ Microsoft.CodeAnalysis.OperationKind.IfStatement = 5 -> Microsoft.CodeAnalysis.O
 Microsoft.CodeAnalysis.OperationKind.IncrementExpression = 518 -> Microsoft.CodeAnalysis.OperationKind
 Microsoft.CodeAnalysis.OperationKind.IndexedPropertyReferenceExpression = 267 -> Microsoft.CodeAnalysis.OperationKind
 Microsoft.CodeAnalysis.OperationKind.InstanceReferenceExpression = 277 -> Microsoft.CodeAnalysis.OperationKind
+Microsoft.CodeAnalysis.OperationKind.InterpolatedStringExpression = 286 -> Microsoft.CodeAnalysis.OperationKind
+Microsoft.CodeAnalysis.OperationKind.InterpolatedStringText = 1037 -> Microsoft.CodeAnalysis.OperationKind
```
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: FYI @jaredpar @gafter for public API change (for IOperation).

**Comment by mavasani** [comment]
Unrelated test seems to be failing in windows_debug_unit64_prtest: Microsoft.CodeAnalysis.CompilerServer.UnitTests.CompilerServerUnitTests.ReferenceCachingCS

@OmarTawfik Is this a known issue? Note this PR is on features/ioperation branch.

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please
// Failing run: https://ci.dot.net/job/dotnet_roslyn/job/features_ioperation/job/windows_debug_unit64_prtest/55/

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by OmarTawfik** [comment]
@mavasani no known issues so far for compiler server tests in master. Did we verify it is not because of changes in this feature branch?

**Comment by mavasani** [comment]
@OmarTawfik There are no changes to the compiler layer that can affect VBCSCompiler, and the test is also passing fine locally. So the failure seems unrelated.

**Comment by mavasani** [comment]
@OmarTawfik Logs here: https://ci.dot.net/job/dotnet_roslyn/job/features_ioperation/job/windows_debug_unit64_prtest/56/artifact/Binaries/Debug/UnitTests/VBCSCompilerTests/xUnitResults/Roslyn.Compilers.CompilerServer.UnitTests.dll.xml

```
19213 System.IO.IOException : The process cannot access the file 'D:\\j\\workspace\\windows_debug---a5b724b3\\Binaries\\Temp\\RoslynTests\\2a56a430-583e-418a-bb24-309d5f6bcb6d\\ReferenceCachingCS\\hello3.exe' because it is being used by another process.
at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath) at System.IO.File.InternalDelete(String path, Boolean checkHost) at Microsoft.CodeAnalysis.Test.Utilities.DisposableFile.Dispose() 
at Microsoft.CodeAnalysis.CompilerServer.UnitTests.CompilerServerUnitTests.<ReferenceCachingCS>d__55.MoveNext() in q:\roslyn\src\Compilers\Server\VBCSCompilerTests\CompilerServerTests.cs:line 867
--- End of stack trace from previous location where exception was thrown
--- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) 
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
--- End of stack trace from previous location where exception was thrown 
--- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
```

Seems like a flaky test. I am going to ignore the failure and file a separate issue to track it.

**Comment by mavasani** [comment]
Filed https://github.com/dotnet/roslyn/issues/19763
